File: HackerRank/ProbolemSolving/Easy/100_Around/29_Weighted_Uniform_Strings.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int getCharCode(char letter) {

    for (int i = 1, j = 97; i < 27; i++, j++) {
        if (letter == j) return i;
    }

    return 0;
}

int getTotal(char *word, int wordLen) {
    int total = 0;

    for (int i = 0; i < wordLen; i++) {
        int letter = word[i],
            charCode = getCharCode(letter);

        total += charCode;
    }

    return total;
}

int *weightedUniformStrings(char *word, int wordLen, int *length) {
    int k = 1;
    int *result = calloc(10000, sizeof(int));
    result[0] = getTotal(word, wordLen);

    for (int i = 0; i < wordLen; i++) {

        if (word[i] == ' ') continue;

        int charCode = getCharCode(word[i]);
        int sum = charCode;

        result[k] = charCode;
        k++;

        for (int j = i + 1; j < wordLen; j++) {
            if (word[j] == ' ') continue;

            if (word[i] == word[j]) {
                sum += charCode;
                result[k] = sum;
                word[j] = ' ';
                k++;
            } else break;
        }

        result[k++] = sum;
    }

    *length = k;

    return result;
}

int isMatch(int number, int *arr, int len) {

    for (int i = 0; i < len; i++) {
        if (number == arr[i]) return 1;
    }

    return 0;
}

int main() {

    char word[100000];

    scanf("%s", word);
    int len = strlen(word), n;

    int resultLen;
    int *result = weightedUniformStrings(word, len, &resultLen);

    scanf("%d", &n);
    int *stringWeight = malloc(n * 4);

    for (int i = 0; i < n; i++) scanf("%d", &stringWeight[i]);

    for (int i = 0; i < n; i++) {
        if(isMatch(stringWeight[i], result, resultLen)) {
            printf("Yes\n");
        } else printf("No\n");
    }

    return 0;
}

File: HackerRank/ProbolemSolving/Easy/100_Around/29_Weighted_Uniform_Strings.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

const charCodes = {}
for (let i = 0, j = 97; i < 26; i++, j++)
    charCodes[String.fromCharCode(j)] = i + 1

const getTotal = string => {
    let total = 0;
    
    for (let letter of string) total += charCodes[letter];

    return total;
}


// Complete the weightedUniformStrings function below.
//# Solved but 2 sample input are wrong 😒
function weightedUniformStrings(s, queries) {
    const string = s.split('');
    const total = getTotal(string);

    const codes =  string.reduce((obj, letter, index) => {
        let charCode = charCodes[letter],
            sum = charCode

        if (charCode) {
            for (let i = index + 1; i < string.length; i++) {
                if (string[i] === letter) {
                    string[i] = ''
                    sum += charCode
                    obj[sum] = sum
                } else break;
            }
            
            obj[charCode] = charCode;
        }

        return obj;

    }, {})

    
    codes[total] = total

    const result = queries.reduce((arr, num) => {
        const find = codes[num]

        if (!find) arr.push('No')
        else arr.push('Yes')

        return arr
    }, [])
    
    return result
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const s = readLine();

    const queriesCount = parseInt(readLine(), 10);

    let queries = [];

    for (let i = 0; i < queriesCount; i++) {
        const queriesItem = parseInt(readLine(), 10);
        queries.push(queriesItem);
    }

    let result = weightedUniformStrings(s, queries);

    ws.write(result.join("\n") + "\n");

    ws.end();
}
